### ISSUE TITLE
We should still process RPC requests even if the consensus failed

### ISSUE THREAD

**Issue by ebuchman** [tendermint/tendermint#1772]
If consensus fails, the node deadlocks. 

We should still be able to eg. process queries.

Eg. https://github.com/cosmos/cosmos-sdk/issues/1216

**Comment by srmo** [comment]
Oh, cool. This is something we sometimes observed too - i.e. RPC requests to status or dumpConsenus no longer being answered and at some point the chain comming to a halt. In 0.19 - 0.20
At this point we are affected by https://github.com/tendermint/tendermint/issues/1754 which under high load scenarios produces the RPC behavior too

**Comment by srmo** [comment]
Oh, cool. This is something we sometimes observed too - i.e. RPC requests to status or dumpConsenus no longer being answered and at some point the chain comming to a halt. In 0.19 - 0.20
At this point we are affected by https://github.com/tendermint/tendermint/issues/1754 which under high load scenarios produces the RPC behavior too

**Comment by melekes** [comment]
> If consensus fails, the node deadlocks.

that's because RPC cannot acquire a lock, right?

**Comment by ebuchman** [comment]
Believe so

**Comment by melekes** [comment]
why can't we just set the timeout <T> on http get request and cancel it if it takes more time than <T>. Almost any HTTP client out there allows to set timeout.

**Comment by melekes** [comment]
1. made `/status` resistant to consensus halt https://github.com/tendermint/tendermint/pull/2009 (validator power is `0` when consensus halts)
2. change consensus state Mutex to RWMutex https://github.com/tendermint/tendermint/pull/2008

don't think we should do the same for `/dump_consensus_state` or `/consensus_state` endpoints cause their entire output depends on consensus state, which we can't get due to mutex.

**Comment by melekes** [comment]
Also, see https://github.com/tendermint/tendermint/issues/2010

**Comment by melekes** [comment]
Merged to develop. Will be shipped with 0.22.5 or next breaking release.

**Comment by ebuchman** [comment]
@melekes do you know how to replicate getting the endpoints to hang with a consensus failure? I tried putting a panic in the DeliverTx but I had no problem with endpoints on v0.22.4 and with localnet-start. Trying to make sure I fully understand the problem and the fix.

**Comment by melekes** [comment]
Oh, I added an extra `cs.mtx.Lock()` to one of the `enterX` functions (forgot which exactly). With this lock, consensus halted and `/status` & `/dump_consensus_state` & `/consensus_state` endpoints became unresponsive.

**Comment by melekes** [comment]
Nope. I haven't tried, there is not enough info. Still, I think my assumption about the lock is correct.

**Comment by melekes** [comment]
Reverted my earlier fix. We should try to figure out why consensus state locks in the first place and only then make a decision. @srmo if you experience the same issue again, please attach an output of `kill -6 <PID>` or steps to reproduce here.

**Comment by ebuchman** [comment]
I wasn't able to replicate the issue by panicing from the state machine but then in doing some review with @xla he suggested that perhaps it has to do with the pubsub - so we should try again to replicate with a panic in the state machine while there are active websocket connections subscribed to blocks. 

On the testnet there were certain active connections and I believe @srmo uses active websockets as well so this is a promising lead!
